src/metaswitch/clearwater/cluster_manager/etcd_synchronizer.py

```diff
@@ -245,31 +244,51 @@ def read_from_etcd(self):
-    # an overall picture of the new cluster.
-    def update_cluster_view(self, cluster_view, new_state):
+    # Write the new cluster view to etcd. We may be expecting to create the key
+    # for the first time.
+    def write_to_etcd(self, old_cluster_view, new_state, with_index=None):
+        cluster_view = old_cluster_view.copy()
+
+        # Update the cluster view based on new state information. If new_state
+        # is a string then it refers to the new state of the local node.
+        # Otherwise, it is an overall picture of the new cluster.
         if isinstance(new_state, str):
             cluster_view[self._ip] = new_state
         elif isinstance(new_state, dict):
             cluster_view = new_state
 
-        return cluster_view
-
-    # Write the new cluster view to etcd. We may be expecting to create the key
-    # for the first time.
-    def write_to_etcd(self, cluster_view):
+        _log.debug("Writing state %s into etcd" % cluster_view)
         json_data = json.dumps(cluster_view)
 
         try:
+            if with_index:
+                self._client.write(self._key, json_data, prevIndex=with_index)
             if self._index is None:
                 self._client.write(self._key, json_data, prevExist=False)
             else:
                 self._client.write(self._key, json_data, prevIndex=self._index)
 
-                # We may have just successfully set the local node to
-                # WAITING_TO_LEAVE, in which case we no longer need the leaving
-                # flag.
-                self._leaving_flag = False
+            # We may have just successfully set the local node to
+            # WAITING_TO_LEAVE, in which case we no longer need the leaving
+            # flag.
+            self._leaving_flag = False
         except ValueError:
-            pass
+            _log.debug("Contention on etcd write")
+            # Our etcd write failed because someone got there before us.
+
+            if isinstance(new_state, str):
+                # We're just trying to update our own state, so it may be safe
+                # to take the new state, update our own state in it, and retry.
+                result = self._client.get(self._key)
+                cluster_view = json.loads(result.value)
+
+                # This isn't safe if someone else has changed our state for us,
+                # or the overall deployment state has changed (in which case we
+                # may want to change our state to something else, so check for
+                # that.
+                if (cluster_view[self._ip] == old_cluster_view[self._ip] and
+                    (self.calculate_cluster_state(cluster_view) ==
+                     self.calculate_cluster_state(old_cluster_view))):
+                    self.write_to_etcd(cluster_view,
+                                       new_state,
+                                       with_index=result.modifiedIndex)
```

Comment:
We may hit a contention on this write too - right? We should at least catch the ValueError. We should also probably do the same logic again, but if we don't I guess we'll just kick the FSM again, which is just a performance hit rather than a bug?